_posts/2018-02-16-running-foreman-with-postgresql-96-from-software-collections.md

```diff
@@ -0,0 +1,90 @@
+
+As the Installer does not have any command line options to configure PostgreSQL, we will have to create a `custom-hiera.yaml` with the required values.
+
+Under the hood, the Installer is using the [puppetlabs-postgresql](https://github.com/puppetlabs/puppetlabs-postgresql/) Puppet module, so that is the source for our customization.
+Initially I wanted to override the `postgresql::server::*` and `postgresql::client::*` values, as documented in the `README.md`, however doing so would require overriding *many* values, as the module generates many variables in `postgresql::params` based on the values from `postgresql::globals`.
+Instead we will override the values directly in `postgresql::globals` and let all the other classes inherit our definitions.
+
+Given that packages from [SoftwareCollections.org](https://www.softwarecollections.org) always come with a prefix for almost everything, we'll have to override the package names, systemd unit names and the data folder.
+
+Luckily, the [PostgreSQL 9.6](https://www.softwarecollections.org/en/scls/rhscl/rh-postgresql96/) software collection also contains `syspaths` packages (`rh-postgresql96-postgresql-syspaths`, `rh-postgresql96-postgresql-server-syspaths`, etc), so that we won't have to worry about scripts expecting `/usr/bin/psql` or similar to exist ("normal" SCL packages don't touch `/usr/bin` and require a `scl enable NAME` to add the SCL-specific `bin` folder to your `PATH`).
+
+Looking at [`postgresql::globals`](https://github.com/puppetlabs/puppetlabs-postgresql/blob/master/manifests/globals.pp) and [`postgresql::params`](https://github.com/puppetlabs/puppetlabs-postgresql/blob/master/manifests/params.pp), we come up with the following settings:
+
+```yaml
+postgresql::params::version: 9.6
+postgresql::globals::client_package_name: rh-postgresql96-postgresql-syspaths
+postgresql::globals::server_package_name: rh-postgresql96-postgresql-server-syspaths
+postgresql::globals::contrib_package_name: rh-postgresql96-postgresql-contrib-syspaths
+postgresql::globals::service_name: rh-postgresql96-postgresql
+postgresql::globals::datadir: /var/opt/rh/rh-postgresql96/lib/pgsql/data
+postgresql::globals::confdir: /var/opt/rh/rh-postgresql96/lib/pgsql/data
+postgresql::globals::bindir: /usr/bin
+```
+
+Just add these lines to `/etc/foreman-installer/custom-hiera.yaml` before running `foreman-installer` and we end up with Foreman running against a PostgreSQL 9.6.
+
+There is one small issue with this setup, though. For some reason `/var/opt/rh/rh-postgresql96/lib/pgsql/data/postgresql.conf` is not getting the correct SELinux context when created by Puppet, and thus PostgreSQL fails to start. Thus you'd need to either run the Installer in permissive mode and relabel the file afterwards (`restorecon -Rv /var/opt/rh/rh-postgresql96/lib/pgsql/data/postgresql.conf`), or expect the failure, relabel and re-run the installier a second time.
```

Comment:
I suspect https://tickets.puppetlabs.com/browse/PUP-2169 is the issue. The tl;dr is that it caches all contexts at the beginning. Probably it doesn't have the context before that, but would install it during the run. If you rerun the installer and it fixes it then you've confirmed it.

Reply:
spot on, this seems to be the issue.